llpc/patch/llpcPatchIntrinsicSimplify.cpp

```diff
@@ -0,0 +1,416 @@
+ * @file  llpcPatchIntrinsicSimplify.cpp
+ * @brief LLPC source file: contains implementation of class Llpc::PatchIntrinsicSimplify.
+ ***********************************************************************************************************************
+ */
+#define DEBUG_TYPE "llpc-patch-intrinsic-simplify"
+
+#if defined(_MSC_VER)
+// Provide M_PI.
+#define _USE_MATH_DEFINES
+#endif
+
+#include <math.h>
+
+#include "llvm/InitializePasses.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+
+#include "llpcPatchIntrinsicSimplify.h"
+#include "llpcTargetInfo.h"
+
+using namespace llvm;
+using namespace Llpc;
+
+namespace Llpc
+{
+
+// =====================================================================================================================
+// Initializes static members.
+char PatchIntrinsicSimplify::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the LLVM pass intrinsic simplifcations.
+FunctionPass* CreatePatchIntrinsicSimplify()
+{
+    return new PatchIntrinsicSimplify();
+}
+
+// =====================================================================================================================
+// Constructor.
+PatchIntrinsicSimplify::PatchIntrinsicSimplify()
+    :
+    FunctionPass(ID),
+    m_pScalarEvolution(nullptr),
+    m_pContext(nullptr),
+    m_pFunction(nullptr)
+{
+    initializePatchIntrinsicSimplifyPass(*PassRegistry::getPassRegistry());
+}
+
+// =====================================================================================================================
+// Get the analysis usage.
+void PatchIntrinsicSimplify::getAnalysisUsage(
+    AnalysisUsage& analysisUsage // [out] The analysis usage for this pass.
+    ) const
+{
+    analysisUsage.addRequired<ScalarEvolutionWrapperPass>();
+    analysisUsage.addPreserved<ScalarEvolutionWrapperPass>();
+    analysisUsage.addRequired<PipelineStateWrapper>();
+    analysisUsage.setPreservesCFG();
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM function.
+bool PatchIntrinsicSimplify::runOnFunction(
+    Function& func) // [in,out] LLVM function to be run on.
+{
+    bool changed = false;
+
+    m_pPipelineState = getAnalysis<PipelineStateWrapper>().GetPipelineState(func.getParent());
+
+    m_pScalarEvolution = &getAnalysis<ScalarEvolutionWrapperPass>().getSE();
+    m_pContext = &func.getContext();
+    m_pFunction = &func;
+
+    for (Function& otherFunc : func.getParent()->functions())
+    {
+        // Skip non intrinsics.
+        if (otherFunc.isIntrinsic() == false)
+        {
+            continue;
+        }
+
+        SmallVector<Value*, 4> usersOfOtherFunc(otherFunc.users());
+
+        for (Value* const pUser : usersOfOtherFunc)
```

Comment:
You don't have to introduce the `usersOfOtherFunc` vector if you just loop over it